File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    answer = 0
    
    # 이동 가능 좌표 = 1, 그 외는 0으로 지도 초기화
    item_map = [[-1] * 102 for _ in range(102)]
    for rec in rectangle:
        # 좌표값 2배하여 ㄷ 케이스 식별
        x1, y1, x2, y2 = map(lambda x: x * 2, rec)
        for i in range(y1, y2 + 1):
            for j in range(x1, x2 + 1):
                if y1 < i < y2 and x1 < j < x2:
                    # 직사각형 내부 = 0
                    item_map[i][j] = 0
                elif item_map[i][j] != 0:
                    # 다른 직사각형의 내부 아니면서 테두리인 경우 = 1
                    item_map[i][j] = 1
    
    dd = [(1, 0), (0, -1), (-1, 0), (0, 1)]
    q = deque()
    q.append((characterX * 2, characterY * 2, 0))
    visited = [[False] * 102 for _ in range(102)]
    visited[characterY*2][characterX*2] = True
    
    while q:
        x, y, dist = q.popleft()
        
        if x == itemX * 2 and y == itemY * 2:
            answer = dist // 2
            break
        
        for dx, dy in dd:
            nx, ny = x + dx, y + dy
            if item_map[ny][nx] == 1 and not visited[ny][nx]:
                q.append((nx, ny, dist + 1))
                visited[ny][nx] = True
    
    return answer

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.cpp
#include <string>
#include <vector>
#include <deque>

using namespace std;

struct Pos {
    int i;
    int j;
    int dist;
};

int di[4] = {0, 0, -1, 1};
int dj[4] = {-1, 1, 0, 0};

int solution(vector<vector<int>> rectangle, int characterX, int characterY, int itemX, int itemY) {
    int answer = 0;
    vector<vector<int>> map(101, vector<int>(101, -1));
    
    // 2배 좌표로 map 설정
    for (auto& rec : rectangle) {
        int x1 = rec[0] * 2, y1 = rec[1] * 2;
        int x2 = rec[2] * 2, y2 = rec[3] * 2;
        
        for (int i = x1; i <= x2; i++) {
            for (int j = y1; j <= y2; j++) {
                if (x1 < i && i < x2 && y1 < j && j < y2) {
                    map[i][j] = 0;
                } else if (map[i][j] != 0) {
                    map[i][j] = 1;
                }
            }
        }
    }
    
    // bfs
    vector<vector<bool>> visited(101, vector<bool>(101, false));
    deque<Pos> q;
    q.push_back({characterX * 2, characterY * 2, 0});
    visited[characterX * 2][characterY * 2] = true;
    
    while (!q.empty()) {
        Pos cur = q.front();
        q.pop_front();
        
        if (cur.i == itemX * 2 && cur.j == itemY * 2) {
            answer = cur.dist / 2;
            break;
        }
        
        for (int i = 0; i < 4; i++) {
            int ni = cur.i + di[i], nj = cur.j + dj[i];
            
            if (2 > ni || ni > 100 || 2 > nj || nj > 100) continue;
            if (!visited[ni][nj] && map[ni][nj] == 1) {
                q.push_back({ni, nj, cur.dist + 1});
                visited[ni][nj] = true;
            }
        }
    }
    
    return answer;
}